File: src/binr.lua
#!/usr/bin/env lua
local help = [[
binr.lua : build rules via stochastic incremental XAI
(c) 2025, Tim Menzies, timm@ieee.org, mit-license.org

Options:
  -h             Show help.
  -e  era=10     Number of rows in an era
  -b  bins=7     Number of bins for discretization.
  -B  Budget=30  Max rows to eval.
  -l  lives=5    Number of lives.
  -r  repeats=20 Number of experimental repeats.
  -s  seed=42    Random number seed.
  -f  file=../data/auto93.csv  ]]

-- coerce(s) --> v ;; Return int or float or bool or string from `s`.
local function coerce(s) 
  if s then return tonumber(s) or s:match'^%s*(.-)%s*$' end end

local the={}; for k,v in help:gmatch("(%S+)=(%S+)") do the[k] = coerce(v) end
math.randomseed(the.seed)

local DATA, NUM, SYM, COLS, clone, adds

--## Lib

local abs,exp,sqrt,log = math.abs, math.exp, math.sqrt, math.log
local floor,min,max,rand,cos = math.floor,math.min,math.max, math.random, math.cos

local say=io.write
local fmt = string.format

-- sort(t,f) --> t ;; Sort `t` using function `f`.
local sort = function(t,f) table.sort(t,f); return t end 

-- lt(f) --> f ;; Return a function that sorts `a` and `b` on `f`.
local lt = function(f) return function(a,b) return f(a) < f(b) end end

-- cat(a) --> s ;; Return a string representation of array `a`.
local cat = function(a) return "{".. table.concat(a," ") .."}" end

-- o(v) --> s ;; Return a string representation of `v`.
local function o(v,     list,dict)
  list=function(a,u) for _,v in ipairs(a) do u[1+#u]=o(v) end; return u end
  dict=function(d,u)
     for k,v in pairs(d) do u[1+#u]=fmt(":%s %s",k,o(v)) end; return sort(u) end
  return type(v) == "number" and fmt(v%1==0 and "%.0f" or "%.3f", v) or
         type(v) ~= "table" and tostring(v) or 
         cat((#v>0 and list or dict)(v,{})) end

-- s2a(s) --> a ;; Return array of words from string `s`, split on ",".
local function s2a(s,   a)
  a={}; for s1 in s:gmatch"([^,]+)" do a[1+#a] = coerce(s1) end; return a end

-- csv(file) --> f ;; Iterator that returns rows from `file`.
local function csv(file,    src)
  src = assert(io.open(file))
  return function(    s)
    s = src:read(); if s then return s2a(s) else src:close() end end end

-- shuffle(t) --> t ;; Randomly shuffle the order of elements in `t`.
local shuffle = function(t,    n)
	for m=#t,2,-1 do n=math.random(m); t[m],t[n]=t[n],t[m] end; return t end

-- cut(a0,n,data) --> t,t ;;Split `a0` at `n` (if `data` exists,split that too).
local function cut(a0,n,  data)
  local a1,a2 = {},{}
  for j,v in ipairs(a0) do if j <= n then a1[1+#a1]=v else a2[1+#a2]=v end end
  return data and clone(data,a1),clone(data,a2) or a1,a2 end

-- mode(d) --> v ;; Return the most frequent key in `d`.
local function mode(d,   v,n)
  v,n = nil,0
  for v1,n1 in pairs(d) do if n1>n then v,n=v1,n1 end end
  return v end 

-- box_muller(mu,sd) --> n ;; Return a random number from a Gaussian `mu`,`sd`.
local function box_muller(mu,sd) 
  return mu + sd * sqrt(-2 * log(rand())) * cos(6.28 * rand()) end

--## Classes

-- DATA(src) --> DATA ;; Create a new DATA, populated with `src`.
function DATA(  src) return adds(src, {n=0,rows={},cols=nil}) end

-- clone(i,src) --> DATA ;; Return a new DATA with same structure as `i`.
function clone(i,  src) return adds(src, DATA{i.cols.names}) end

-- NUM(at,s) --> NUM ;; Create a NUM object to summarize numbers.
function NUM(at,s) 
  return {at=at or 0, of=s, n=0, mu=0, m2=0, sd=0, bins={},
          best=(tostring(s) or ""):find"+$" and 1 or 0} end

-- SYM(at,s) --> SYM ;; Create a SYM object to summarize symbols.
function SYM(at,s) return {at=at, of=s, n=0, has={}, bins={}} end

-- COLS(row) --> COLS ;; Create a COLS object from a list of column names.
function COLS(row,    t,x,y,all)
  x,y,all = {},{},{}
  for n,s in ipairs(row) do
    all[n] = (s:match"^[A-Z]" and NUM or SYM)(n,s)
    if not s:match"X$" then
      t = s:find"[+-]$" and y or x
      t[1+#t] = all[n] end end 
  return {all=all, x=x, y=y, names=row} end

--## Methods

-- add(i,v) --> v ;; Update `i` with `v`.
local function add(i,v)
  if v == "?" then return v end
  i.n = i.n + 1
  if i.has then i.has[v] = 1 + (i.has[v] or 0) 
  elseif i.mu then 
    local d = v - i.mu
    i.mu = i.mu + d / i.n
    i.m2 = i.m2 + d * (v - i.mu)
    i.sd = i.n<2 and 0 or sqrt((max(0,i.m2)/(i.n - 1))) 
  elseif i.rows then  
    if not i.cols then i.cols = COLS(v) else
      for _,col in pairs(i.cols.all) do add(col, v[col.at]) end 
      i.rows[1 + #i.rows] = v end end
  return v end  

-- adds(src,it) --> it ;; Update `it` with all items from `src`.
function adds(src, it)
  it = it or NUM()
  if type(src) == "string" 
  then for row in csv(src) do add(it,row) end
  else for _,row in pairs(src or {}) do add(it,row) end end
  return it end

-- norm(num,v) --> n ;; Normalize `v` 0..1 using `i`.
local function norm(num,v)
  return  1 / (1 + math.exp(-1.7 * (v - num.mu)/(num.sd + 1e-32))) end

-- bin(i,v) --> n ;; Normalize `v` 0..bins-1 using `i`.
local function bin(i,v)
  return (i.has or v=="?") and v or floor( the.bins * norm(i,v)) end

-- disty(data,row) --> n ;; Return distance of `row` to best goal (using Y cols).
local function disty(data,row,     d)
  d=0; for _,y in pairs(data.cols.y) do d= d + (norm(y, row[y.at]) - y.best)^2 end
  return sqrt(d/#data.cols.y)  end

local function distys(data,rows)
  return sort(rows or data.rows,
              function(a,b) return disty(data,a) < disty(data,b) end) end

--## Think

local function scoreGet(data,row,    b,n)
  n = 0
  for _,col in pairs(data.cols.x) do
    b = bin(col, row[col.at])
    if b ~= "?" then 
      if col.bins[b] then
         n = n + col.bins[b].mu  end end end
  return n end
      
local function scorePut(data,row,     b,y)
  y = disty(data,row)
  for _,col in pairs(data.cols.x) do
    b = bin(col, row[col.at]) 
    if b ~= "?" then
      col.bins[b] = col.bins[b] or NUM(col.at, b)
      add(col.bins[b], y) end end end

local function scoreGuess(data,m,n,rows,    t)
  t = {} 
  --print((m or 1),min(#rows, n or #rows))
  for n = (m or 1),min(#rows, n or #rows) do 
    if n <= #rows then
      t[1+#t] = {scoreGet(data, rows[n]), rows[n]} end end
  return sort(t, function(a,b) return a[1] < b[1] end) end

local function scoresSeen(data,      t,m,eps)
  t={}; for m,row in pairs(data.rows) do t[1+#t] = disty(data,row) end
  t=sort(t)
  m=#t//10
  eps = 0.35 * (t[9*m] - t[m])/2.56 
  print(fmt("%.2f, %.2f, %.2f, %.2f, %.2f; eps= %.2f", 
                  t[m], t[3*m], t[5*m], t[7*m], t[9*m], eps)) 
  return data,eps end

local function score(data,eps,     labelled,rows,bestRow,besty,loves,best,y,lives,n)
  print""
  labelled = clone(data)
  besty = 1e32
  lives = lives or the.lives
  seen  = {}
  n=0;
  for m,row in pairs(data.rows) do
    if lives < 0 or n >= the.Budget then break end  
    add(labelled, row)
    scorePut(labelled, row)
    seen[row]=row; n=n+1
    if m % the.era==0 then 
      best = scoreGuess(labelled, 1, m+20, data.rows)[1][2]
      if not seen[best] then seen[best]=best; n=n+1 end
      y = disty(data, best)
      if y < besty - eps
      then besty,bestRow = y,best  ; say"!"
      else lives = lives - 1       ; say"."
      end end end 
  return bestRow, besty,n end 

--## Demos

local egs={}

egs["-h"] = function(_) print("\n"..help.."\n") end
egs["-s"] = function(n) math.randomseed(n or the.seed); the.seed =n end
egs["--the"] = function(_) print(o(the)) end
egs["--shuffle"] = function(_) print(o(shuffle{10,20,30,40,50})) end

egs["--csv"] = function(_,    n) 
  n=1; for row in csv(the.file) do
         if n % 25 == 0 then print(o(row)) end 
         n = n + 1 end end

egs["--num"] = function(_,num)
  num=NUM()
  for _=1,1000 do add(num, box_muller(10,5)) end
  print(fmt("%.3f %.3f", num.mu, num.sd)) end 

egs["--data"] = function(_) 
  for n,col in pairs(DATA(the.file).cols.x) do 
  print(n,o(col)) end end

egs["--disty"]= function(_,    data,num,t)
  data,t = DATA(the.file), {}
  for n,row in pairs(data.rows) do
    if n % 25 == 0 then t[1+#t] = disty(data,row) end end 
  print(o(sort(t))) end 

egs["--score"] = function(_,    t,data,eps,y) 
                   data,eps = scoresSeen(DATA(the.file)) 
                   t={}
                   for n = 1,the.repeats do
                     data.rows  = shuffle(data.rows)
                     _,y,seen = score(data,eps) 
                     t[n] = 100*y//1 end
                   print("\n"..o(sort(t))) end

egs["--all"] = function(_,   n) 
                n = the.seed
                for k,_ in pairs(egs) do 
                  math.randomseed(n)
                  if k~="--all" then print("\n-----",k); egs[k]() end end end

-- cli(d,funs) --> nil ;; Update `d` with flags from command-line; run `funs`.
local function cli(d,funs)
  for i,s in pairs(arg) do
    if funs[s]
    then funs[s](coerce(arg[i+1]))
    else for k,_ in pairs(d) do
          if k:sub(1,1)==s:sub(2) then d[k]=coerce(arg[i+1]) end end end end end

if arg[0]:find"binr.lua" then cli(the,egs) end

File: src/binr.py
#!/usr/bin/env python3 -B
# vim: ts=2:sw=2:sts=2:et
"""
binr.py : build rules via stochastic incremental XAI   
(c) 2025, Tim Menzies, timm@ieee.org, mit-license.org

Options:

    -h             Show help.  
    -b  bins=4     Number of bins for discretization (int).  
    -B  Budget=30  Max rows to eval (int).  
    -C  CF=0.8     crossover rate  
    -F  F=0.3      scale factor between two nums.  
    -e  era=10     Number of rows in an era (int)  
    -p  p=2        Distance coefficient  
    -r  repeats=20 Number of experimental repeats (int).  
    -s  seed=42    Random number seed (int).  
    -f  file=../data/auto93.csv File to load (str).
"""
from math import floor,sqrt,cos,log,exp,pi
from typing import Any,Iterable
import fileinput,random,sys,re
rand = random.random

class obj(dict):
  "Structs with slots accessiable via x.slot. And pretty print." 
  def __repr__(i): return "{" + ' '.join(f":{k} {o(i[k])}" for k in i) + "}"
  def __setattr__(i, k, v): i[k] = v
  def __getattr__(i, k): 
    try: return i[k]
    except KeyError: raise AttributeError(k)

the = obj(bins=4, Budget=30, CF=.8, era=10, F=0.3, p=2, repeats=20, seed=42,
          file="../data/auto93.csv")

# types, upper case
QTY  = float | int
ATOM = QTY | str | bool
ROW  = list[ATOM]
ROWS = list[ROW]
NUM,SYM,TRI, COLS = obj, obj,obj,obj 
COL  = NUM | SYM  # not TRO                   
COLS = list[list[COL]]
DATA = tuple[ROWS, COLS]             

# ------------------------------------------------------------------------------
# Constructors, mixed case
def Sym(has:dist=None) -> SYM: 
  "Summarize symbol."
  return obj(it=Sym, n=0, has=has or {}, bins={})

def Num(mu=0, sd=1) -> NUM:
  "Summarize numbers."
  return obj(it=Num, n=0, mu=mu, sd=sd, m2=0, bins={})

def Tri(lo=0, mid=0.5, hi=1) -> TRI:  
  "Used to sample from a skewed distribution but (sub/adding not defined)."
  return obj(it=Tri,n=0, lo=lo, mid=mid, hi=hi)

def Col(at=0, of=" ") -> COL:
  "Column in rows of data."
  it = (Num if of[0].isupper() else Sym)()
  it.at = at
  it.of = of
  it.best = str(of)[-1]!="-" 
  return it

def Cols(names:list[str]) -> COLS:
  "Factory. Turns column names into columns."
  cols = [Col(at=i, of=s) for i,s in enumerate(names)]
  return obj(it=Cols, names=names, 
           all = cols,
           x   = [col for col in cols if str(col.of)[-1] not in "+-X"],
           y   = [col for col in cols if str(col.of)[-1] in "+-"])

def Data(rows = None) -> DATA:
  "Summarize rows into columns."
  return adds(rows, obj(it=Data, n=0, rows=[], cols=None))

def clone(data:DATA, rows=None) -> DATA:
  "Mimic the structure of `data`. Optinally, add some rows."
  return adds(rows, Data([data.cols.names]))

# ------------------------------------------------------------------------------
def add(i: NUM | SYM | DATA,   # NOTE: TRI not supported (cant decrement lo,hi) 
        item: Any,
        inc = 1) -> Any: # returns item
  "Add or subtract items from columns or data."
  if item=="?": return item
  i.n += inc
  if   i.it is Sym: i.has[item] = inc + i.has.get(item,0)
  elif i.it is Num:
    item = float(item)
    if inc < 0 and i.n < 2:
      i.n = i.mu = i.sd = i.m2 = 0
    else:
      d     = item - i.mu
      i.mu += inc * d / i.n
      i.m2 += inc * d * (item - i.mu)
      i.sd  = 0 if i.n < 2 else sqrt(max(0,i.m2)/(i.n - 1))
  elif i.it is Data:
    if i.cols: 
      row = [add(c, item[c.at], inc) for c in i.cols.all] 
      i.rows.append(row) if inc > 0 else i.rows.remove(row)
    else: i.cols = Cols(item)
  return item

def sub(i,item): 
  "Subtract items."
  return add(i,item,-1)

def adds(items:Iterable = None, it=None ) -> obj: # returns it
  "Load many items into `it` (defeault is `Num()`)."
  it = it or Num()
  if str(items)[-4:]==".csv":
    with open(items, encoding="utf-8") as f:
      for line in f:
        if line: add(it, [s.strip() for s in line.split(",")])
  else: [add(it, item) for item in (items or [])]
  return it

# ------------------------------------------------------------------------------
def sample(i: TRI | SYM | NUM | list) -> list:
  "Sample a value from a TRI / Num / Sym / Data summary."
  if type(i)==list: return [sample(col) for col in i]
  if i.it is Num: return irwinHall3(i.mu, i.sd)
  if i.it is Tri:
    p = (i.mid - i.lo) / (i.hi - i.lo + 1e-32)
    u, v = rand(), rand()
    return i.lo + (i.hi - i.lo) * (min(u, v) + p * abs(u - v))
  if i.it is Sym:
    r = rand() * i.n
    for x, count in i.has.items():
      r -= count
      if r <= 0: return x
    return x # should never get here.

def mixtures(data: list[COL], np=100) -> Data:
  "Return `n` samples nonparametrically: add the delta between two items to a third."
  any = lambda: random.choice(data.rows)
  return [mixture(data, any(), any(), any()) for _ in range(np)]

def mixture(data:DATA, a:ROW, b:ROW, c:ROW) -> ROW:
  "Mutate `a` by mixing items from `b,c`."
  def nump(z): return type(z) in [float,int]
  d = a[:]
  keep = random.randrange(len(a))
  for j,(A,B,C,col) in enumerate(zip(a,b,c,data.cols.all)):
    if j != keep and rand() < the.CF:
      d[j] = B if rand() < 0.5 else C
      if col.it is Num and nump(A) and nump(B) and nump(C): 
        d[j] = wrap(col, A + the.F*(B - C))
  return d

def wrap(num,v):
  "Restrain `v` to the effective min,max range of `num`." 
  lo,hi = num.mu - 3*num.sd, num.mu + 3*num.sd
  if v<lo: return hi - ((lo-v) % (hi-lo))
  if v>hi: return lo + ((v-hi) % (hi-lo))
  return v

# ------------------------------------------------------------------------------
def mid(i: COL | DATA) -> ATOM | ROW: 
  "Return the expected value of `i`."
  if i.it is Num: return i.mu
  if i.it is Tri: return i.mid
  if i.it is Sym: return max(i.has, key=i.has.get)
  return [mid(col) for col in i.cols.all]

def shuffle(lst:list) -> list:
  "Shuffle `lst` in place."
  random.shuffle(lst); return lst

def irwinHall3(mu=0,sd=1) -> float:
  "Fast normal sampling: chatgpt.com/share/6935eb44-705c-8010-8782-454c0aff8a5e"
  return mu + sd * 2.0 * (rand() + rand() + rand() - 1.5)

def marsagliaPolar(mu=0,sd=1) -> float: 
  "Slightly slower normal sampling."
  while 1:
     u,v = 2*rand()-1, 2*rand()-1
     s = u*u + v*v
     if 0 < s < 1: return mu + sd*u*sqrt(-2*log(s)/s)

def norm(i:NUM, v:QTY) -> float: 
  "Returns 0..1."
  return 1/(1+exp(-1.7*(v-i.mu)/(i.sd+1e-32))) if i.it is Num and v!="?" else v 

def bin(col:COL, v:ATOM) -> int | ATOM:
  "Returns 0..bins-1."
  return floor(the.bins * norm(col,v) ) if col.it is Num and v!="?" else v 

def dist(src:Iterable) -> float:
  "Mankoski distance."
  d,n = 0,0
  for z in src: 
    n += 1
    d += z ** the.p
  return (d/n) ** (1/the.p)

def disty(data:DATA, row:ROW) -> float:
  "Distance of `row` to `best` values in each goal column."
  return dist(abs(norm(col, row[col.at]) - col.best) for col in data.cols.y)

def distx(data:DATA, row1:ROW, row2:ROW) -> float:
  "Distance between `x` attributes of two rows."
  return dist(_aha(col, row1[col.at], row2[col.at]) for col in data.cols.x)

def _aha(col:COL, a:ATOM, b:ATOM) -> float:
  "If any unknowns, assume max distance."
  if a==b=="?": return 1
  if col.it is Sym : return a != b
  a,b = norm(col,a), norm(col,b)
  a = a if a != "?" else (0 if b>0.5 else 1)
  b = b if b != "?" else (0 if a>0.5 else 1)
  return abs(a - b)

# ------------------------------------------------------------------------------
def scoreGet(model, use, row:ROW) -> ROW:
  "Sum the score of the bins used by `row`."
  n = 0
  for slot in use:
    if (v := row[slot.at]) != "?":
      if bin(model.cols.all[slot.at], v) == slot.of:
        n += want(slot)
  return n

def scorePut(data:DATA, row:ROW, score:QTY):
  "Increment the bins used by `row`."
  for x in data.cols.x:
    if (b := bin(x, row[x.at])) != "?":
      one = x.bins[b] = x.bins.get(b) or Num()
      one.at, one.of = x.at, b
      add(one, score)

def want(slot): return slot.mu  + slot.sd/sqrt(slot.n)

def top(data):
  return sorted((slot for x in data.cols.x for slot in x.bins.values()),key=want)

dump={}

def score(data:DATA, eps=0.05):
  "Guess next few scores using scores seen to date."
  best_score, best_row = 1e32, None
  rows = shuffle(data.rows)
  seen, model = set(), Data([data.cols.names])
  for j, row in enumerate(rows):
    if len(seen) >= the.Budget: break
    add(model, row) 
    scorePut(model, row, disty(model, row))
    seen.add(id(row))
    if (j+1) % the.era == 0 and j < len(rows) - 100:
      use = top(model)
      use = use[-int(sqrt(len(use))):]
      for slot in use:
          k=(slot.at, slot.of)
          dump[k] = 1 + dump.get(k,0)
      candidate = min(rows[j+1:j+100], key=lambda r: scoreGet(model,use, r))
      seen.add(id(candidate))
      if (score := disty(model, candidate)) < best_score - eps:
        best_score, best_row = score, candidate
  return best_score

# -----------------------------------------------------------------------------
def o(x):
  "Pretty print."
  if type(x) is type(o) : return x.__name__ + '()'
  if type(x) is float : return str(int(x)) if x == int(x) else f"{x:,.2f}"
  if type(x) is list : return "["+(', '.join(o(y) for y in x))+"]"
  return str(x)

# ------------------------------------------------------------------------------
def go_h(_) -> None: 
  print(__doc__)

def go__the(_) -> None: 
  print(the)

def go_s(n: str) -> None: 
  the.seed = float(n); random.seed(the.seed)

def go__sym(_) -> None:  
  print(adds("aaaabbc",Sym()))

def go__num(_) -> None:
  print(adds(irwinHall3(10,2) for _ in range(10**3)))

def go__data(f = None) -> None:
  data = Data(f or the.file)
  print(data.cols.x[-1])
  print(len(data.rows),data.rows[1])

def go__disty(f = None):
  ys, data = Num(), Data(f or the.file)
  print(*[col.of for col in data.cols.all],"y",sep="\t")
  Y=lambda row: floor(100*disty(data,row))
  for r in sorted(data.rows,key=Y)[::20]:
    print(*r,Y(r),sep="\t")

def go__distx(f = None):
  xs, data = Num(), Data(f or the.file)
  print(*[col.of for col in data.cols.all],"x",sep="\t")
  X=lambda row1: floor(100*distx(data,row1, data.rows[0]))
  for r in sorted(data.rows,key=X)[::20]:
    print(*r,X(r),sep="\t")

def go__inc(f=None):
  data1 = Data(f or the.file)
  data2 = clone(data1)
  for row in data1.rows:
    add(data2,row)
    if len(data2.rows)==50: print(o(mid(data2)))
  print(o(mid(data2)))
  for row in data1.rows[::-1]:
    if len(data2.rows)==50: print(o(mid(data2)))
    sub(data2,row)

def f(x)    : return 1.61 + 2.1*x[0] - 3.5*(x[1]*2) + 4*(x[2]**3) - 5*(x[3]**4)
def fx(row) : print(obj(best=row, y=f(row)))

def go__random(_):
  eden = [Num(100,10), Num(20,5), Num(10,4), Num(3,2)]
  fx( min((sample(eden) for _ in range(1000)), key=f))

def go__hclimb(_):
  m,r   = 100,9
  model = [("X1",100,10), ("X2",20,5), ("X3",10,4), ("X4",3,2)]
  eden  = [Num(mu,sd) for _,mu,sd in model]
  data  = Data([[s for s,_,_ in model]] + [sample(eden) for _ in range(m)])
  for _ in range(r):
    tmp = clone(data, sorted(data.rows, key=f)[:m//2])
    fx(tmp.rows[0])
    data = clone(data, mixtures(tmp,m))

def go__score(f= None):
  my   = lambda n: floor(100*n)
  data = Data(f or the.file)
  print(len(data.rows))
  ys   = adds(my(disty(data,row)) for row in data.rows)
  print(obj(mu=ys.mu,sd=ys.sd))
  print(*sorted(my(score(data)) for _ in range(the.repeats)))
  print(sorted((n,k) for k,n in dump.items()))

_tests= {k:fun for k,fun in vars().items() if "go__" in k}

def go__all(_):
  for k,fun in _tests.items(): 
    if k != "go_all": print("\n----- "+k); random.seed(the.seed); fun(_)

# ------------------------------------------------------------------------------
if __name__ == "__main__":
  for n, s in enumerate(sys.argv):
    if fn := vars().get(f"go{s.replace('-', '_')}"): 
      random.seed(the.seed)
      fn(sys.argv[n+1] if n < len(sys.argv)-1 else None)
